### PR TITLE
Adding bulk inserts or updates for PostgreSQL

### PR DESCRIPTION
**Pull request by PedroD**
Adding support for PostgreSQL bulk insert or update (also allowing to bulk ignore conflicts).

```kotlin
         database.bulkInsert(Employees) {
                item {
                    set(it.id, 1)
                    set(it.name, "vince")
                    set(it.job, "engineer")
                    set(it.salary, 1000)
                    set(it.hireDate, LocalDate.now())
                    set(it.departmentId, 1)
                }
                item {
                    set(it.id, 5)
                    set(it.name, "vince")
                    set(it.job, "engineer")
                    set(it.salary, 1000)
                    set(it.hireDate, LocalDate.now())
                    set(it.departmentId, 1)
                }

                onDuplicateKey(Employees.id) {
                        set(it.salary, it.salary + 900)
                }
            }
```
(Will add +900 to each already existing employee salary)

```kotlin
         database.bulkInsert(Employees) {
                item {
                    set(it.id, 1)
                    set(it.name, "vince")
                    set(it.job, "engineer")
                    set(it.salary, 1000)
                    set(it.hireDate, LocalDate.now())
                    set(it.departmentId, 1)
                }
                item {
                    set(it.id, 5)
                    set(it.name, "vince")
                    set(it.job, "engineer")
                    set(it.salary, 1000)
                    set(it.hireDate, LocalDate.now())
                    set(it.departmentId, 1)
                }

                onDuplicateKey(Employees.id) {
                }
            }
```
(Will ignore any already existing employee and continue the query)

### CODE REVIEW

File: ktorm-support-postgresql/src/main/kotlin/org/ktorm/support/postgresql/BulkInsertOrUpdate.kt [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+ *              set(it.departmentId, 1)
+ *          }
+ *
+ *          onDuplicateKey(Employees.id) {
+ *              set(it.salary, it.salary + 900)
+ *          }
+ *      }
+ * ```
+ *
+ * @since 2.7
```
Comment (PedroD) [inline ktorm-support-postgresql/src/main/kotlin/org/ktorm/support/postgresql/BulkInsertOrUpdate.kt:87]: @vincentlauvlwj what version should go here?

**Comment by vincentlauvlwj** [comment]
Thank you for your contribution!!

One question: What's the difference between calling `onDuplicateKey` with an empty closure and not calling it? What if a user wants to throw an exception when any key conflict exists? Seems it is a more common case than just ignoring conflicts.

**Comment by vincentlauvlwj** [comment]
Also, please update the `build.gradle` file, add your GitHub ID to the developers info (line 90), that will let more people know your contributions.

**Comment by PedroD** [comment]
> Thank you for your contribution!!
> 
> One question: What's the difference between calling `onDuplicateKey` with an empty closure and not calling it? What if a user wants to throw an exception when any key conflict exists? Seems it is a more common case than just ignoring conflicts.

Good question, I haven't though about that scenario.

But think about it now, I only see two possibilities:

Given that this method is a Insert or Update, it will potentially never throw an error on finding duplicates with the given key columns. And it should be like this by design of PostgreSQL. So the developer using a query like:

```SQL
INSERT INTO $table ($col1, $col2, ...)
VALUES (...), (...), (...), ...
ON CONFLICT ($col1, $col2, ...)
DO UPDATE SET ....
```
Must be aware that this will never return an error for the first-level "collision" of ($col1, $col2, ...).

So, my design line of though was the following: Given that when I use "upserts" I either want to:
1. Insert as new record if not exists, or update the existing record matching the conflict columns I mention in the query
2. Insert as new record if not exists, and ignore record if already exists matching the conflict columns I mention in the query
3. Insert as new record if not exists, and receive an error if it conflicts

So, to account for scenario 1 we can use the form:
```kotlin
// Generates a DO UPDATE SET
onDuplicateKey(Employees.id) {
      set(it.salary, it.salary + 900)
}
```

For scenario 2 we can use the form:
```kotlin
// Generates a DO NOTHING
onDuplicateKey(Employees.id) {
}
```
For scenario 3, it cannot be done in this type of query as far as I understand by design of PostgreSQL. There's no way to throw an error if a record already exists matching the conflict columns I mention in the query.
However, it will throw an error from the DBMS if:
* it finds a collision with any column constraint not matching the conflict ones mentioned in the query
* if during the DO UPDATE operation one or more SETs create a new collision

This is all behavior that is native from PostgreSQL, given that are only two actions we can do in these type of queries (NOTHING and UPDATE) [according to their documentation](https://www.postgresql.org/docs/9.5/sql-insert.html).

So, for scenario 3, the user should probably use the default `batchInsert` which, by not performing any update/upsert, will return an error on collisions.

Does this make sense to you?

**Comment by PedroD** [comment]
PS. I made some commits renaming these methods to make clear what they do (Insert or Update).

Also, what is the difference between a `batchInsert` and a `bulkInsert` in the case of MySQL?
Should I use `bulkInsertOrUpdate` or `batchInsertOrUpdate` in the method I just implemented on this PR?

**Comment by vincentlauvlwj** [comment]
Thank you for the explanation, I fully understand what you mean. But would it be possible to generate an insert SQL without `on conflict` clause when `onDuplicateKey` is not called? For example: 

```kotlin
database.bulkInsert(Employees) {
    item {
        set(it.name, "jerry")
        set(it.job, "trainee")
        set(it.managerId, 1)
        set(it.hireDate, LocalDate.now())
        set(it.salary, 50)
        set(it.departmentId, 1)
    }
    item {
        set(it.name, "linda")
        set(it.job, "assistant")
        set(it.managerId, 3)
        set(it.hireDate, LocalDate.now())
        set(it.salary, 100)
        set(it.departmentId, 2)
    }
}
```

Then generated SQL:

```sql
INSERT INTO t_employee (name, job, manager_id, hire_date, salary, department_id) VALUES (?, ?, ?, ?, ?, ?), (?, ?, ?, ?, ?, ?) 
```

This is exactly the scenario 3 you mentioned, it inserts new records if not exists, and throws an error on collisions.

**Comment by vincentlauvlwj** [comment]
If it is possible to make `onDuplicateKey` optional, then we don't need to rename the function to `bulkInsertOrUpdate`. `bulkInsert` is good.

**Comment by PedroD** [comment]
> If it is possible to make `onDuplicateKey` optional, then we don't need to rename the function to `bulkInsertOrUpdate`. `bulkInsert` is good.

Ok, I get the idea, I didn't go that way because it didn't seem to be the way the code was designed initially, meaning that I will need to change this behavior to allow an empty "conflictTarget" to exist:

image

**Comment by vincentlauvlwj** [comment]
Merged, I will release v3.3.0 these days.

**Comment by vincentlauvlwj** [comment]
`batchInsert` is implemented based on JDBC's `addBatch` and `executeBatch`, it generates multiple SQL statements, while `bulkInsert` generates only one SQL statement, that's the difference.

**Comment by vincentlauvlwj** [comment]
You are right, the name `bulkInsert` is not good enough, with this name it should not have the Update part. 

Maybe it is better to have another function `bulkInsertOrUpdate`, I will do a refactoring based on your code.

**Comment by PedroD** [comment]
@vincentlauvlwj awesome!

Keep up the great work! :muscle:

**Pull request merged**